src/pages/messages/messages.ts

```diff
@@ -57,40 +59,135 @@ export class MessagesPage implements OnInit, OnDestroy {
   }
 
   ngOnInit() {
-    MeteorObservable.subscribe('messages', this.selectedChat._id).subscribe(() => {
-      MeteorObservable.autorun().subscribe(() => {
-        this.messages = Messages.find(
-          {chatId: this.selectedChat._id},
-          {sort: {createdAt: 1}}
-        ).map((messages: Message[]) => {
-          messages.forEach((message: Message) => {
-            message.ownership = this.senderId == message.senderId ? 'mine' : 'other';
-          });
-
-          return messages;
-        });
-      });
-    });
-
     this.autoScroller = this.autoScroll();
+    this.subscribeMessages();
+
+    // Get total messages count in database so we can have an indication of when to
+    // stop the auto-subscriber
+    MeteorObservable.call('countMessages').subscribe((messagesCount: number) => {
+      // Note that the 'scroller' element is being created dynamically by an Ionic
+      // component and therefore the event listener can't be registered directly
+      // from the view
+      let scrollListener = this.onScroll.bind(this);
+      this.scroller.addEventListener('scroll', scrollListener);
+      this.autoRemoveScrollListener(messagesCount, scrollListener);
+    });
   }
 
   ngOnDestroy() {
     this.autoScroller.disconnect();
   }
 
-  onInputKeypress({keyCode}: KeyboardEvent): void {
+  onInputKeypress({ keyCode }: KeyboardEvent): void {
     if (keyCode == 13) {
       this.sendMessage();
     }
   }
 
+  onScroll(e): void {
+    // Unless we're at the top of the messages page
+    if (!this.scroller.scrollTop) {
+      this.subscribeMessages();
+    }
+  }
+
+  showOptions(): void {
+    const popover = this.popoverCtrl.create(MessagesOptionsComponent, {
+      chat: this.selectedChat
+    }, {
+      // Will be used as a CSS selector in our style-sheet
+      cssClass: 'options-popover'
+    });
+
+    popover.present();
+  }
+
+  // Subscribes to the relevant set of messages
+  subscribeMessages(): Subscription {
+    // Prohibit parallel subscriptions
+    if (this.loadingMessages) return;
+    // A flag which indicates if there's a subscription in process
+    this.loadingMessages = true;
+    // A custom offset to be used to re-adjust the scrolling position once
+    // new dataset is fetched
+    this.scrollOffset = this.scroller.scrollHeight;
+
+    return MeteorObservable.subscribe('messages',
```

Comment:
The `loadingMessages` boolean is used like some sort of a mutex where it prevents from parallel subscription requests to be done all at the same time which can potentially overflow the Meteor server with unnecessary requests.